File: dashboard/lib/restart.py
"""The Restart verb — a LOCAL COMMAND execution, the dashboard's SECOND button class (issue #116),
a sibling of Tidy (``lib/tidy.py``).

Every GitHub-write button lives in ``lib/actions.py``. Restart, like Tidy, is deliberately
different: it shells the local ``superlooper`` CLI — here ``superlooper request-restart`` — to ask
the LIVE runner to restart ITSELF in its own cmux tab. The runner honors the request between ticks
by re-exec'ing in place: a fresh process image that reloads the currently-installed engine and
clears in-memory episode state (e.g. a tripped systemic-launch hold). This is "turn the loop off
and on again" as one tap instead of finding the runner's tab, Ctrl-C, and retyping ``superlooper
run``.

The bright line that shapes the whole design: **the button never spawns or places a cmux tab**
(owner ruling, 2026-07-09 — automated tab placement is out). It only asks a runner that is ALREADY
running in its own tab to re-exec there. So:

* **This adapter never raises into a caller.** A missing binary, a timeout, a killed process — all
  become a nonzero rc + empty stdout (mirrors ``lib/tidy._run`` / ``lib/gh._run``), so a tap can
  only ever fail closed.
* **The dead-runner case is an HONEST outcome, not a crash.** No live runner ⇒ the CLI exits
  nonzero but prints a well-formed JSON body (``running: false`` + the one-line manual start). So
  unlike Tidy (which treats any nonzero rc as an error), this adapter parses the CLI's JSON FIRST —
  a refusal is a truthful result the button shows plainly ("no loop running"), never a generic
  error, and never a launch or placement attempt. Only when stdout carries NO parseable object does
  the adapter fall back to an rc-based error (the CLI is missing or truly crashed).
* **A WATCHED repo only.** Every invocation is gated on an allow-list mapping each configured repo
  slug to its checkout path; a stray/forged request for an unwatched repo is refused BEFORE any
  subprocess runs (the same bright line ``Actions`` and ``Tidy`` draw).

The CLI to run is the CONFIGURED path (config's ``superlooper_cli``), but ``SL_SUPERLOOPER`` overrides
it — exactly so ``tests/conftest.py`` can point every test at an absent binary by default and a
restart test can inject the fake in-body. This mirrors ``lib/tidy``'s ``SL_SUPERLOOPER`` precedence.
"""
import json
import os
import subprocess

# Per-call hard timeout (seconds). A module constant, not a literal, so a test can shrink it and
# trip the timeout path (mirrors tidy._DEFAULT_TIMEOUT).
_DEFAULT_TIMEOUT = 30


def _binary(configured):
    """The superlooper CLI to run: the ``SL_SUPERLOOPER`` env override wins over the configured path
    (config's ``superlooper_cli``), mirroring ``lib/tidy``'s precedence so the entry point and the
    tests agree on binary resolution."""
    return os.environ.get("SL_SUPERLOOPER") or configured


def _run(binary, args, timeout=None):
    """Run ``<binary> <args>``; returns ``(rc, stdout, stderr)``. NEVER raises: a timeout, a missing
    binary, or any OSError is caught and returned as a nonzero rc with empty stdout so the caller
    fails closed (mirrors ``tidy._run``)."""
    if timeout is None:
        timeout = _DEFAULT_TIMEOUT
    try:
        proc = subprocess.run([binary, *args], capture_output=True, text=True, timeout=timeout)
        return proc.returncode, proc.stdout, proc.stderr
    except subprocess.TimeoutExpired:
        return 124, "", "timed out"            # conventional timeout rc
    except (OSError, ValueError):
        return 127, "", "command not found"    # missing binary / bad invocation


def parse_result(stdout):
    """The single JSON object ``superlooper request-restart --json`` prints, or ``None`` when stdout
    carries no parseable object (a missing/crashed CLI). Pure and unit-tested, so the coupling to the
    CLI's ``--json`` contract is pinned by a test rather than discovered in production."""
    txt = (stdout or "").strip()
    if not txt:
        return None
    try:
        val = json.loads(txt)
    except (ValueError, TypeError):
        return None
    return val if isinstance(val, dict) else None


def _error(rc, stderr, binary):
    """A plain, honest failure message for a CLI that didn't answer — what the UI shows instead of a
    fake success. Names the CLI on a missing binary so the operator knows exactly what to fix."""
    stderr = (stderr or "").strip()
    if rc == 127:
        return ("could not run the superlooper CLI at %s — is it installed? "
                "(set 'superlooper_cli' in config.json)" % binary)
    if rc == 124:
        return "superlooper request-restart timed out"
    return stderr or ("superlooper request-restart failed (exit %d)" % rc)


class Restart:
    """The Restart verb, bound to the configured superlooper CLI path, an allow-list mapping each
    WATCHED repo slug to its checkout path, and the operator display name it signs the request with
    (issue #58 — recorded in the marker, journaled by the runner). Two methods back the button's
    two-step flow: :meth:`preflight` reports whether a live runner exists and writes NOTHING (the
    dialog decides what to show), :meth:`execute` drops the request (only after the in-UI confirm).
    Every result is honest — ``running``/``ok`` are the real command outcome, never a pretend one."""

    def __init__(self, binary, repo_paths, operator=None, timeout=None):
        self._binary = binary
        self._paths = dict(repo_paths or {})
        self._operator = operator if (isinstance(operator, str) and operator.strip()) else None
        self._timeout = timeout

    def _refuse(self, verb):
        # An unwatched repo is refused BEFORE any subprocess — the command runner only ever targets
        # the checkouts the operator configured (bright line: never steerable off-machine/off-repo).
        return {"ok": False, "verb": verb, "running": None, "error": "unknown repo"}

    def preflight(self, repo):
        """Report whether a live runner exists for ``repo`` — runs ``request-restart --check`` (which
        writes NOTHING). ``running`` is the honest liveness (``None`` if the CLI couldn't answer);
        ``manual`` carries the one-line manual start the dead-runner dialog shows."""
        path = self._paths.get(repo)
        if path is None:
            return self._refuse("restart-check")
        return self._invoke("restart-check", path, ["--check"])

    def execute(self, repo):
        """Ask the live runner to restart itself — runs ``request-restart`` (drops the marker; the
        in-UI confirm already happened), signed with the operator + a ``command-center`` source. On a
        runner that died since the preflight, the CLI's honest refusal (``running: false`` + the
        manual start) is surfaced as-is — never a launch attempt, never a generic error."""
        path = self._paths.get(repo)
        if path is None:
            return self._refuse("restart")
        extra = ["--source", "command-center"]
        if self._operator:
            extra += ["--operator", self._operator]
        return self._invoke("restart", path, extra)

    def _invoke(self, verb, path, extra):
        binary = _binary(self._binary)
        rc, out, err = _run(binary, ["request-restart", "--repo", path, "--json", *extra],
                            timeout=self._timeout)
        parsed = parse_result(out)
        if parsed is not None:
            # The CLI answered (even a dead-runner refusal is a well-formed body at rc 1) — surface
            # its honest outcome, normalizing the verb name for the UI trail.
            parsed["verb"] = verb
            return parsed
        # No parseable JSON ⇒ the CLI is missing or crashed: a plain, honest failure (never a false
        # "runner up", never a silent success).
        return {"ok": False, "verb": verb, "running": None, "error": _error(rc, err, binary)}

File: dashboard/static/restart.js
/* The Restart button's confirm dialog (issue #116) — the dashboard's SECOND ops-verb button and a
   sibling of Tidy: tapping Restart asks the LIVE runner (via the server → `superlooper
   request-restart`) to restart ITSELF in its own cmux tab. It is NOT a GitHub write — it drops a
   request the runner honors between ticks by re-exec'ing in place (fresh engine, cleared in-memory
   state). So the flow is deliberately two-step and confirm-gated (tap-where-you-read, design §0.3):

     open → POST /api/restart/check → dialog states EXACTLY what will happen (or, if NO runner is
     live, says so and shows the one-line manual start) → the owner taps "Restart the loop" →
     POST /api/restart → the honest result is shown.

   The bright line the whole design rides on: this NEVER spawns or places a cmux tab (owner ruling,
   2026-07-09). With no live runner the dialog cannot and must not resurrect one — it reports that
   plainly and points at the manual procedure; there is no confirm button, because there is nothing
   to ask. A command failure (missing CLI, crash) comes back as ok:false with an error string and is
   shown plainly — never a silent success.

   window.CCRestart is a persistent overlay OUTSIDE #root (like the drawer/tidy) so the 2s poll never
   touches it. Design B.1: this file computes NO semantics — the server already turned the CLI's JSON
   into a structured result; this only binds it to pixels. */
(function () {
  "use strict";

  function esc(s) {
    return String(s == null ? "" : s)
      .replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;");
  }
  function el(id) { return document.getElementById(id); }

  // `listedRepo` is the repo the dialog is CURRENTLY showing (set only when its preflight renders);
  // the confirm executes against THAT, never the mutable `slug`, so a re-open can't leave the dialog
  // showing repo A while confirm restarts repo B. `gen` supersedes an in-flight preflight when a
  // newer open starts, so an out-of-order response is dropped.
  var node = null, slug = "", listedRepo = "", busy = false, gen = 0;

  function postJSON(path, payload) {
    return fetch(path, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload)
    }).then(function (r) {
      return r.json().catch(function () { return {}; }).then(function (b) {
        return { status: r.status, ok: r.ok, body: b || {} };
      });
    });
  }

  function ensure() {
    if (node) return;
    node = document.createElement("div");
    node.id = "cc-restart";
    node.className = "cc-restart";
    node.innerHTML =
      '<div class="cc-restart-card">' +
        '<div class="cc-restart-head">' +
          '<span class="cc-restart-title">\u{1F504} RESTART <b id="cc-restart-target"></b></span>' +
          '<span class="cc-restart-sub">asks the running loop to restart itself in its own tab — ' +
            'runs <code>superlooper request-restart</code> on this machine. no GitHub, no AI.</span>' +
          '<button class="cc-restart-x" data-restart-close title="close (Esc)">✕</button>' +
        '</div>' +
        '<div class="cc-restart-body" id="cc-restart-body"></div>' +
      '</div>';
    document.body.appendChild(node);

    node.addEventListener("click", function (e) {
      var t = e.target;
      if (!t || !t.closest) return;
      if (t === node || t.closest("[data-restart-close]") || t.closest("[data-restart-cancel]")) {
        close();
        return;
      }
      if (t.closest("[data-restart-confirm]")) { runExecute(); return; }
      var retryEl = t.closest("[data-restart-retry]");
      if (retryEl) { loadPreflight(++gen, retryEl.getAttribute("data-restart-retry")); return; }
    });
    document.addEventListener("keydown", function (e) {
      if (isOpen() && e.key === "Escape" && !busy) close();
    });
  }

  function isOpen() { return !!(node && node.classList.contains("open")); }

  function open(repoSlug) {
    if (!repoSlug) return;
    ensure();
    slug = repoSlug;
    listedRepo = "";                 // nothing confirmed-ready yet for this open
    busy = false;
    el("cc-restart-target").textContent = "→ " + slug;
    node.classList.add("open");
    loadPreflight(++gen, repoSlug);  // ++gen supersedes any preflight still in flight from a prior open
  }

  function close() { if (node) node.classList.remove("open"); }

  // Step 1 — the preflight: is a live runner there to ask? (writes nothing). Each preflight carries
  // the generation of the open (or retry) that started it and the repo it is checking; a response
  // from a superseded open, or one that arrives after a close, is dropped.
  function loadPreflight(myGen, repo) {
    setBody('<div class="cc-restart-loading">checking for a running loop…</div>');
    postJSON("/api/restart/check", { repo: repo })
      .then(function (res) {
        if (myGen !== gen || !isOpen()) return;    // superseded / closed → ignore this stale reply
        var b = res.body || {};
        if (res.status !== 200) { renderError(b.error, false, repo); return; }
        if (b.running === true) { listedRepo = repo; renderConfirm(); return; }
        if (b.running === false) { renderNoRunner(b.manual); return; }
        renderError(b.error, false, repo);         // running unknown → the CLI couldn't answer
      })
      .catch(function () {
        if (myGen === gen && isOpen()) renderError("couldn’t reach the command center", false, repo);
      });
  }

  // Step 2 — the confirmed request: ask the runner to restart. Reached ONLY from the
  // data-restart-confirm tap, and it targets the EXACT repo the dialog is showing (listedRepo).
  function runExecute() {
    if (busy || !listedRepo) return;
    var repo = listedRepo, myGen = gen;
    busy = true;
    setBody('<div class="cc-restart-loading">asking the loop to restart…</div>');
    postJSON("/api/restart", { repo: repo })
      .then(function (res) {
        busy = false;
        if (myGen !== gen || !isOpen()) return;    // a re-open superseded this / dialog closed
        var b = res.body || {};
        if (res.status !== 200) { renderError(b.error, true, repo); return; }
        if (b.requested === true) { renderDone(); return; }
        // The runner died between the preflight and the confirm → honest no-runner, never an error.
        if (b.running === false) { renderNoRunner(b.manual); return; }
        renderError(b.error, true, repo);
      })
      .catch(function () {
        busy = false;
        if (myGen === gen && isOpen()) renderError("couldn’t reach the command center", true, repo);
      });
  }

  // A live runner is there: state EXACTLY what will happen, in plain words, then the confirm gate.
  function renderConfirm() {
    setBody(
      '<div class="cc-restart-lead">A loop is running for <b>' + esc(slug) + '</b>. Restarting will:</div>' +
      '<ul class="cc-restart-consequence">' +
        '<li>let it <b>finish the current tick</b>, then restart the loop in its own cmux tab;</li>' +
        '<li>reload the currently-installed engine and <b>clear in-memory state</b> ' +
          '(e.g. a stuck launch hold);</li>' +
        '<li>leave <b>in-flight worker sessions untouched</b> — nothing merges while it restarts.</li>' +
      '</ul>' +
      '<div class="cc-restart-actions">' +
        '<button class="btn ghost" data-restart-cancel>Cancel</button>' +
        '<button class="btn primary" data-restart-confirm>Restart the loop</button>' +
      '</div>');
  }

  // No live runner: the button cannot (and must not) resurrect one. Say so plainly and show the
  // one-line manual start — never a confirm button, because there is nothing to ask.
  function renderNoRunner(manual) {
    var line = manual || "open a cmux tab and run: superlooper run --repo <path>";
    setBody(
      '<div class="cc-restart-notice">No loop is running for <b>' + esc(slug) + '</b> — ' +
        'there is nothing to restart. Start it by hand in a visible cmux tab:</div>' +
      '<div class="cc-restart-manual"><code>' + esc(line) + '</code></div>' +
      '<div class="cc-restart-actions"><button class="btn ghost" data-restart-close>Done</button></div>');
  }

  function renderDone() {
    setBody(
      '<div class="cc-restart-result ok">✓ Restart requested — the loop will finish its ' +
        'current tick and restart itself in its own tab.</div>' +
      '<div class="cc-restart-actions"><button class="btn ghost" data-restart-close>Done</button></div>');
  }

  // A command failure is never a silent success: show the honest error. When the PREFLIGHT failed
  // (nothing was asked), offer a Retry; when the EXECUTE failed, just a dismiss.
  function renderError(message, wasExecute, repo) {
    var msg = message || (wasExecute ? "restart failed — nothing was asked" : "couldn’t check the loop");
    var retry = (!wasExecute && repo)
      ? '<button class="btn ghost" data-restart-retry="' + esc(repo) + '">Retry</button>' : "";
    setBody(
      '<div class="cc-restart-result err">⚠ ' + esc(msg) + '</div>' +
      '<div class="cc-restart-actions">' + retry +
        '<button class="btn ghost" data-restart-close>Close</button></div>');
  }

  function setBody(html) {
    var b = el("cc-restart-body");
    if (b) b.innerHTML = html;
  }

  window.CCRestart = { open: open, isOpen: isOpen };
})();
